File: java/src/test/java/sa/amnn/kit/WebhookTest.java
package sa.amnn.kit;

import static org.junit.jupiter.api.Assertions.*;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;
import sa.amnn.kit.webhooks.WebhookHandler;
import sa.amnn.kit.webhooks.WebhookSignature;

class WebhookTest {
    static final String SECRET = "unit-test-secret";

    // Golden vector captured from a real Amen delivery (2026-08-30).
    static final String GV_SECRET = "whsec_test";
    static final String GV_TS = "2026-08-30T18:53:23.885957+00:00";
    static final String GV_BODY = "{\"event\":\"deal.status.changed\",\"timestamp\":\"2026-08-30T18:53:23.885957+00:00\",\"payload\":{\"number\":\"D-0000000002\",\"status\":\"paid\"}}";
    static final String GV_HEX = "950ca0ff7494dd435d4dc9d7e7ebe31cf54f0859a28a69a686d77e8db9dfd45c";

    @Test void goldenVector() {
        byte[] body = GV_BODY.getBytes(StandardCharsets.UTF_8);
        assertEquals("sha256=" + GV_HEX, WebhookSignature.compute(GV_SECRET, GV_TS, body));
        assertTrue(WebhookSignature.verify(GV_SECRET, GV_TS, body, "sha256=" + GV_HEX));
        assertTrue(WebhookSignature.verify(GV_SECRET, GV_TS, body, GV_HEX)); // prefix optional
    }

    @Test void signatureRoundtripAndTamper() {
        String ts = "2026-08-30T18:53:23.885957+00:00";
        byte[] body = "{\"event\":\"deal.status.changed\",\"timestamp\":\"" + ts + "\"}".getBytes(StandardCharsets.UTF_8);
        String sig = WebhookSignature.compute(SECRET, ts, body);
        assertTrue(sig.startsWith("sha256="));
        assertTrue(WebhookSignature.verify(SECRET, ts, body, sig));
        // tampered body
        assertFalse(WebhookSignature.verify(SECRET, ts, (new String(body, StandardCharsets.UTF_8) + " ").getBytes(StandardCharsets.UTF_8), sig));
        // tampered timestamp (signature is bound to it)
        assertFalse(WebhookSignature.verify(SECRET, "2026-08-30T18:53:23.885958+00:00", body, sig));
        // missing signature / missing timestamp
        assertFalse(WebhookSignature.verify(SECRET, ts, body, null));
        assertFalse(WebhookSignature.verify(SECRET, null, body, sig));
        assertFalse(WebhookSignature.verify(SECRET, "", body, sig));
    }

    @Test void handlerVerifiesDedupesAndDispatches() {
        List<String> seen = new ArrayList<>();
        WebhookHandler h = new WebhookHandler(SECRET, e -> seen.add(e.type() + ":" + e.id()));
        String ts = "2026-08-30T18:53:23.885957+00:00";
        byte[] body = ("{\"event\":\"deal.status.changed\",\"timestamp\":\"" + ts + "\",\"payload\":{\"status\":\"paid\"}}").getBytes(StandardCharsets.UTF_8);
        Map<String, String> good = new HashMap<>();
        good.put(WebhookSignature.SIGNATURE_HEADER, WebhookSignature.compute(SECRET, ts, body));
        good.put(WebhookSignature.TIMESTAMP_HEADER, ts);
        good.put(WebhookSignature.EVENT_HEADER, "deal.status.changed");

        // bad signature -> 401
        Map<String, String> bad = new HashMap<>(good);
        bad.put(WebhookSignature.SIGNATURE_HEADER, "sha256=bad");
        assertEquals(401, h.handle(bad, body).status());

        // missing timestamp -> 401
        Map<String, String> noTs = new HashMap<>(good);
        noTs.remove(WebhookSignature.TIMESTAMP_HEADER);
        assertEquals(401, h.handle(noTs, body).status());

        // case-insensitive header lookup, first delivery -> 200
        Map<String, String> lower = new HashMap<>();
        lower.put("x-webhook-signature", WebhookSignature.compute(SECRET, ts, body));
        lower.put("x-webhook-timestamp", ts);
        WebhookHandler.Result first = h.handle(lower, body);
        assertEquals(200, first.status());
        assertEquals(Map.of("ok", true), first.body());

        // duplicate (same event timestamp) -> 200 + duplicate flag, not re-dispatched
        WebhookHandler.Result dup = h.handle(good, body);
        assertEquals(200, dup.status());
        assertEquals(true, dup.body().get("duplicate"));

        assertEquals(List.of("deal.status.changed:" + ts), seen);
    }
}

File: kotlin/amen-client/src/test/kotlin/sa/amnn/kit/WebhookTest.kt
package sa.amnn.kit

import kotlinx.coroutines.test.runTest
import sa.amnn.kit.webhooks.WebhookHandler
import sa.amnn.kit.webhooks.WebhookSignature
import kotlin.test.*

class WebhookTest {
    private val secret = "unit-test-secret"

    // Golden vector captured from a real Amen delivery (verified 2026-08-30).
    private val goldenSecret = "whsec_test"
    private val goldenTs = "2026-08-30T18:53:23.885957+00:00"
    private val goldenBody = """{"event":"deal.status.changed","timestamp":"2026-08-30T18:53:23.885957+00:00","payload":{"number":"D-0000000002","status":"paid"}}"""
    private val goldenHex = "950ca0ff7494dd435d4dc9d7e7ebe31cf54f0859a28a69a686d77e8db9dfd45c"

    @Test fun goldenVector() {
        val computed = WebhookSignature.compute(goldenSecret, goldenTs, goldenBody.toByteArray())
        assertEquals(goldenHex, computed)
        assertTrue(WebhookSignature.verify(goldenSecret, goldenTs, goldenBody.toByteArray(), computed))
        assertTrue(WebhookSignature.verify(goldenSecret, goldenTs, goldenBody.toByteArray(), "sha256=$computed"))
    }

    @Test fun signatureRoundtrip() {
        val ts = "2026-08-30T00:00:00+00:00"
        val body = """{"event":"deal.status.changed","timestamp":"$ts"}""".toByteArray()
        val sig = WebhookSignature.compute(secret, ts, body)
        assertTrue(WebhookSignature.verify(secret, ts, body, sig))
        assertTrue(WebhookSignature.verify(secret, ts, body, "sha256=$sig"))
        // tampered body must fail
        assertFalse(WebhookSignature.verify(secret, ts, body + ' '.code.toByte(), sig))
        // tampered timestamp must fail (timestamp is part of the signed message)
        assertFalse(WebhookSignature.verify(secret, ts + "x", body, sig))
        // missing signature / missing timestamp must fail
        assertFalse(WebhookSignature.verify(secret, ts, body, null))
        assertFalse(WebhookSignature.verify(secret, null, body, sig))
    }

    @Test fun handlerVerifiesDispatchesAndDedupes() = runTest {
        val seen = mutableListOf<String>()
        val types = mutableListOf<String?>()
        val h = WebhookHandler(secret) { seen += it.id; types += it.type }
        val ts = "2026-08-30T18:53:23.885957+00:00"
        val body = """{"event":"deal.status.changed","timestamp":"$ts","payload":{"number":"D-1","status":"paid"}}""".toByteArray()
        val sig = WebhookSignature.compute(secret, ts, body)
        val good = mapOf(
            WebhookSignature.SIGNATURE_HEADER to "sha256=$sig",
            WebhookSignature.TIMESTAMP_HEADER to ts,
            WebhookSignature.EVENT_HEADER to "deal.status.changed",
        )

        // bad signature -> 401
        assertEquals(401, h.handle(good + (WebhookSignature.SIGNATURE_HEADER to "sha256=bad"), body).status)
        // missing timestamp header -> 401
        assertEquals(401, h.handle(mapOf(WebhookSignature.SIGNATURE_HEADER to "sha256=$sig"), body).status)
        // valid -> 200 ok, dispatched once, deduped on the event timestamp
        assertEquals(mapOf("ok" to true), h.handle(good, body).body)
        assertEquals(true, h.handle(good, body).body["duplicate"])
        assertEquals(listOf(ts), seen)
        assertEquals(listOf<String?>("deal.status.changed"), types)
    }
}
